pkg/compactor/bucket_compactor.go

```diff
@@ -361,7 +362,12 @@ func (c *BucketCompactor) runCompactionJob(ctx context.Context, job *Job) (shoul
 	uploadBegin := time.Now()
 	uploadedBlocks := atomic.NewInt64(0)
 
+	if err = verifyCompactedBlocksTimeRanges(ctx, compIDs, toCompactMinTime.UnixMilli(), toCompactMaxTime.UnixMilli(), c.blockSyncConcurrency, subDir); err != nil {
+		return false, nil, err
```

Comment:
I'm a bit scared of failing the compaction in a first stage. There's always the risk that we're missing something and there may be legit edge cases where the check will fail but it's not an issue. I'm wondering if we should start smoother: if the check fails, we log it, we track a metric, we alert on that metric, but we move forward with the compaction. Then, after some time running in prod, based on what we learned we'll decide whether we should make it an hard check or not.

Reply:
I agree with what Marco said: Let's first test this with warning+metric on our systems, and double-check how it works on real data.